discovery/manager.go

```diff
@@ -0,0 +1,288 @@
+//
+// Discoverer does not know if an actual change happened.
+// It does guarantee that it sends the new TargetGroup whenever a change happens.
+//
+// Discoverers should initially send a full set of all discoverable TargetGroups.
+type Discoverer interface {
+	// Run hands a channel to the discovery provider(consul,dns etc) through which it can send
+	// updated target groups.
+	// Must returns if the context gets canceled. It should not close the update
+	// channel on returning.
+	Run(ctx context.Context, up chan<- []*config.TargetGroup)
+}
+
+// type pool struct {
+// 	cancel func()
+// 	tgroups []*config.TargetGroup
+// }
+
+// NewManager is the Discovery Manager constructor
+func NewManager(ctx context.Context, logger log.Logger) *Manager {
+	return &Manager{
+		ctx:            ctx,
+		logger:         logger,
+		actionCh:       make(chan func()),
+		syncCh:         make(chan map[string][]*config.TargetGroup),
+		targets:        make(map[string]map[string][]*config.TargetGroup),
+		discoverCancel: []context.CancelFunc{},
+	}
+}
+
+// Manager maintains a set of discovery providers and sends each update to a channel used by other packages.
+type Manager struct {
+	ctx            context.Context
+	logger         log.Logger
+	syncCh         chan map[string][]*config.TargetGroup // map[targetSetName]
+	actionCh       chan func()
+	discoverCancel []context.CancelFunc
+	targets        map[string]map[string][]*config.TargetGroup // map[targetSetName]map[providerName]
+}
+
+// Run starts the background processing
+func (m *Manager) Run() error {
+	for {
+		select {
+		case f := <-m.actionCh:
+			f()
+		case <-m.ctx.Done():
+			m.cancelDiscoverers()
+			return m.ctx.Err()
+		}
+	}
+}
+
+// SyncCh returns a read only channel used by all Discoverers to send target updates.
+func (m *Manager) SyncCh() <-chan map[string][]*config.TargetGroup {
+	return m.syncCh
+}
+
+// ApplyConfig removes all running discovery providers and starts new ones using the provided config.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	err := make(chan error)
+	m.actionCh <- func() {
+		m.cancelDiscoverers()
+		for _, scfg := range cfg.ScrapeConfigs {
+			for provName, prov := range m.providersFromConfig(scfg.ServiceDiscoveryConfig) {
+				m.startProvider(scfg.JobName, provName, prov)
+			}
+		}
+		close(err)
+	}
+
+	return <-err
+}
+
+func (m *Manager) startProvider(jobName, provName string, worker Discoverer) {
+	ctx, cancel := context.WithCancel(m.ctx)
+	updates := make(chan []*config.TargetGroup)
+
+	m.discoverCancel = append(m.discoverCancel, cancel)
+
+	go worker.Run(ctx, updates)
+	go func(provName string) {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case tgs, ok := <-updates:
+				// Handle the case that a target provider exits and closes the channel
+				// before the context is done.
+				if !ok {
+					return
+				}
+				m.syncCh <- m.mergeGroups(jobName, provName, tgs)
+			}
+		}
+	}(provName)
```

Comment:
This is not in a loop so there's no reason to pass this as an argument.

Reply:
I'd in fact rename the function to `runProvider` and have it block within that for loop. Then use `go m.runProvider(...)` at the call site.
Starting  a producer go routine is fine and obviously necessary but completely detaching work of a method often just makes it harder to reuse in other contexts (mostly testing).